### PR TITLE
[WIP] Add inline transformations of several event types

### PR DESCRIPTION
**Pull request by tim-cools**
See:https://github.com/JasperFx/marten/issues/404

I changed the api a bit, and you can also configure it from the initialize like this

```
 _.Events.Project()
    .Event<QuestStarted>((session, streamId, @event) => doSomethingWithTheEvent(@event))
```

Tests are running from IDE. I Couldn't run the global RAKE script due to dotnet versioncore issues on this dev machine. Will fix this soon, but let me know if you have problems running the tests on TC.

### CODE REVIEW

**Comment by CoreyKaylor** [comment]
Updated the title just to indicate that it's not ready to pull yet.

**Comment by jeremydmiller** [comment]
@tim-cools I'm sorry, but could I talk you into rebasing this on master before I pull it in?

**Comment by daveaglick** [comment]
This looks awesome and is exactly what I need right now.

I do have a comment/question: should the `Project()` and `InlineTransformation()` methods be part of the `ProjectionCollection` instead of the `EventGraph` (which might change the method names slightly)? That would match the convention used by other inline projections like aggregates.
```
options.Events.InlineProjections.Project()
    .Event<QuestStarted>((session, streamId, evt) => events.Add(evt))
    .Event<MembersJoined>((session, streamId, evt) => events.Add(evt))
    .Event<MonsterSlayed>((session, streamId, evt) => events.Add(evt));
// renamed from InlineTransformation() to match AggregateStreamsWith<T>()
options.Events.InlineProjections.TransformStreamsWith(projection);
```

This actually bring up a broader point (which probably needs it's own issue or something) - as a new user I find the API around event projections to be a bit confusing. Especially with regards to the differences around aggregates, projections, and transformations. These terms seem to be used somewhat interchangeably, both in the core class/interface definitions and in the supporting registration methods. It's not clear from the outset why something might be an aggregate over a projection. Is one a superset of the other? As an example, I add an *aggregate* through a class called `InlineProjections`. These kinds of questions could probably be addressed with some kind of glossary in the help documents around projections or alternatively by taking a look at the API prior to 1.0 and refining some of the semantics.

**Comment by daveaglick** [comment]
Just realized that `InlineTransformation()` isn't actually part of this PR and was already there. The comment about moving it under the `ProjectionCollection` still stands but might be out of scope for this PR. I think that illustrates the second point about the API though. Why do I add an inline `IProjection` projection under the root `EventGraph` but then add an aggregate through a property called `InlineProjections`? The public event projection API just seems a little jumbled in it's current state.

**Comment by tim-cools** [comment]
@jeremydmiller @daveaglick Thanks for the feedback! I'm out of town but will look into this in the beginning of next week...

**Comment by jeremydmiller** [comment]
@tim-cools I looked closer at this one, and I'm honestly a little dubious. Having to specify a separate lambda for synchronous vs. asynchronous invocations is going to be a pretty serious drag.

**Comment by tim-cools** [comment]
@jeremydmiller I updated the PR. 

I also felt that *Having to specify a separate lambda for synchronous vs. asynchronous invocations is going to be a pretty serious drag.*

So I updated the projection to be more intelligent. It now also fetches the projected view from the db. This is what I do in each projection anyway. Doing so the projection implementation only needs to update the projected view data. The (a)sync is only used to to load the projected view from db, so a single configuration can be used for both sync an async.

```
            StoreOptions(_ =>
            {
                _.AutoCreateSchemaObjects = AutoCreate.All;
                _.Events.ProjectView<PersistedView>()
                    .ProjectEvent<QuestStarted>((view, @event) => events.Add(@event))
                    .ProjectEvent<MembersJoined>(e => e.QuestId, (view, @event) => events.Add(@event))
                    .ProjectEvent<MonsterSlayed>(e => e.QuestId, (view, @event) => events.Add(@event));
            });
```

And like this if you want to use a class:

```

    public class PersistDocumentProjection : ViewProjection<PersistedView>
    {
        public PersistDocumentProjection()
        {
            ProjectEvent<QuestStarted>(Persist);
            ProjectEvent<MembersJoined>(e => e.QuestId, Persist);
            ProjectEvent<MonsterSlayed>(e => e.QuestId, Persist);
        }

        private void Persist<T>(PersistedView view, T @event)
        {
            view.Events.Add(@event);
        }
    }
```

Let me know if I broke some else as some not related tests are failing on my dev machine...

@daveaglick I noticed you had a discussion about the projection but had no time to get into the details. Let me know if I missed something important?

**Comment by daveaglick** [comment]
@tim-cools Thanks for the heads up! I had actually starting working on my own implementation based on this one as a starting point to play around with some edge cases I encountered while applying Marten to some LoB apps. Things like views based on events with multiple levels of indirection to the aggregate. I'm out of the office today, but I'll take a look tomorrow and see if the updates hit these use cases (or how I can help enhance the PR of not).

**Comment by daveaglick** [comment]
This is going to be a long one...sorry! I want to make sure I've properly documented the *why* and *how* of what I've been working on.

I've been playing with aggregate projection support for a few days, starting out with your PR, but expanding a bit to encompass the following requirements:
* Projections should support both one-to-one and many-to-one event-to-view mappings (I didn't look at many-to-many since I was focused on aggregate projections).
* A projection should be able to specify a view object based on each individual event.
* Specifying the view object should be optional, and if one isn't specified a default behavior such as loading the view from the event stream ID should be used.

These are requirements that I actually have. Particularly the generation of view projections from multiple events that have several levels of indirection to the "main" aggregate root object. I ran into these types of scenarios pretty quickly when experimenting with porting a complex line of business application.

An example that might go with the whole "quest" domain is to consider that each `Party` has `Person` members, each `Person` can carry one or more named `Weapon`, and that there are also `Tavern` objects where a `Party` can dine. Now suppose you want to create a read-side view projection from various events listing the name of all of the weapons present at a given tavern. A `Weapon` belongs (references) a `Person`, which in turn belongs to a `Party`, which dines at the `Tavern`. If I process an event like `WeaponRenamed`, I'll need to follow the associations all the way up that chain to construct my view projection at event processing time.

@jeremydmiller suggested I take a look at using the existing support for this, possibly by switching out the `AggregateFinder<T>` implementation in a `AggregationProjection<T>`. The main challenge with this is that `AggregateFinder<T>` assumes the aggregate view is the same for all the events in a given stream (I.e., it takes the view ID from the stream ID).

Anyway, here's what I came up with. This is from LINQPad so it can be executed as a test. The actual `Projection<TView>` class in near the bottom.

<details>
<summary>IProjection implementation and example/test code</summary>
```
void Main()
{
	DocumentStore store = DocumentStore.For(x =>
	{
		x.Connection("host=localhost;database=marten_test;password=postgres;username=postgres");
		x.AutoCreateSchemaObjects = AutoCreate.All;
		x.Logger(new NulloMartenLogger());


		// Set up events
		x.Events.DatabaseSchemaName = "events";
		x.Events.AddEventType(typeof(AAdded));
		x.Events.AddEventType(typeof(BAdded));
		x.Events.AddEventType(typeof(CAdded));
		x.Events.AddEventType(typeof(CNameChanged));
		x.Events.InlineProjections.AggregateStreamsWith<A>();
		x.Events.InlineProjections.AggregateStreamsWith<B>();
		x.Events.InlineProjections.AggregateStreamsWith<C>();
		x.Events.InlineProjections.Add(new CNamesByAProjection());
	});
	store.Advanced.Clean.CompletelyRemoveAll();
	store.Advanced.Clean.DeleteAllDocuments();

	// Create some As
	Guid a1Id = Guid.NewGuid();
	$"a1: {a1Id}".Dump();
	Guid a2Id = Guid.NewGuid();
	$"a2: {a2Id}".Dump();
	using (var session = store.OpenSession())
	{
		session.Events.StartStream<A>(a1Id, new AAdded
		{
			Id = a1Id,
			Name = "A1"
		});
		session.Events.StartStream<A>(a2Id, new AAdded
		{
			Id = a2Id,
			Name = "A2"
		});
		session.SaveChanges();
	}

	// Create some Bs
	Guid a1b1Id = Guid.NewGuid();
	$"a1b1: {a1b1Id}".Dump();
	Guid a1b2Id = Guid.NewGuid();
	$"a1b2: {a1b2Id}".Dump();
	Guid a2b1Id = Guid.NewGuid();
	$"a2b1: {a2b1Id}".Dump();
	using (var session = store.OpenSession())
	{
		session.Events.StartStream<B>(a1b1Id, new BAdded
		{
			Id = a1b1Id,
			AId = a1Id
		});
		session.Events.StartStream<B>(a1b2Id, new BAdded
		{
			Id = a1b2Id,
			AId = a1Id
		});
		session.Events.StartStream<B>(a2b1Id, new BAdded
		{
			Id = a2b1Id,
			AId = a2Id
		});
		session.SaveChanges();
	}

	// Create some Cs
	Guid a1b1c1Id = Guid.NewGuid(); // Blue
	$"a1b1c1: {a1b1c1Id}".Dump();
	Guid a1b1c2Id = Guid.NewGuid(); // Red
	$"a1b1c2: {a1b1c2Id}".Dump();
	Guid a1b2c1Id = Guid.NewGuid(); // Green
	$"a1b2c1: {a1b2c1Id}".Dump();
	Guid a2b1c1Id = Guid.NewGuid(); // Yellow
	$"a2b1c1: {a2b1c1Id}".Dump();
	Guid a2b1c2Id = Guid.NewGuid(); // Purple
	$"a2b1c2: {a2b1c2Id}".Dump();
	using (var session = store.OpenSession())
	{
		session.Events.StartStream<C>(a1b1c1Id, new CAdded
		{
			Id = a1b1c1Id,
			BId = a1b1Id,
			Name = "Blue"
		});
		session.Events.StartStream<C>(a1b1c2Id, new CAdded
		{
			Id = a1b1c2Id,
			BId = a1b1Id,
			Name = "Red"
		});
		session.Events.StartStream<C>(a1b2c1Id, new CAdded
		{
			Id = a1b2c1Id,
			BId = a1b2Id,
			Name = "Green"
		});
		session.Events.StartStream<C>(a2b1c1Id, new CAdded
		{
			Id = a2b1c1Id,
			BId = a2b1Id,
			Name = "Yellow"
		});
		session.Events.StartStream<C>(a2b1c2Id, new CAdded
		{
			Id = a2b1c2Id,
			BId = a2b1Id,
			Name = "Purple"
		});
		session.SaveChanges();
	}

	// Get the current projection
	using (var session = store.OpenSession())
	{
		CNamesByA[] results = session.Query<CNamesByA>().ToArray();
		Debug.Assert(results.First(x => x.Id == a1Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a1b1c1Id, "Blue")));
		Debug.Assert(results.First(x => x.Id == a1Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a1b1c2Id, "Red")));
		Debug.Assert(results.First(x => x.Id == a1Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a1b2c1Id, "Green")));
		Debug.Assert(results.First(x => x.Id == a2Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a2b1c1Id, "Yellow")));
		Debug.Assert(results.First(x => x.Id == a2Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a2b1c2Id, "Purple")));
		results.Dump();
	}

	// Change a couple names
	using (var session = store.OpenSession())
	{
		session.Events.Append(a1b1c2Id, new CNameChanged
		{
			Name = "Red2"
		});
		session.Events.Append(a2b1c1Id, new CNameChanged
		{
			Name = "Yellow2"
		});
		session.SaveChanges();
	}

	// Get the current projection
	using (var session = store.OpenSession())
	{
		CNamesByA[] results = session.Query<CNamesByA>().ToArray();
		Debug.Assert(results.First(x => x.Id == a1Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a1b1c1Id, "Blue")));
		Debug.Assert(results.First(x => x.Id == a1Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a1b1c2Id, "Red2")));
		Debug.Assert(results.First(x => x.Id == a1Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a1b2c1Id, "Green")));
		Debug.Assert(results.First(x => x.Id == a2Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a2b1c1Id, "Yellow2")));
		Debug.Assert(results.First(x => x.Id == a2Id).CNamesByC
			.Contains(new KeyValuePair<Guid, string>(a2b1c2Id, "Purple")));
		results.Dump();
	}
}


// Aggregates

public class A
{
	public Guid Id { get; set; }
	public string Name { get; set; }

	public void Apply(AAdded added)
	{
		Id = added.Id;
		Name = added.Name;
	}
}

public class B
{
	public Guid Id { get; set; }
	public Guid AId { get; set; }

	public void Apply(BAdded added)
	{
		Id = added.Id;
		AId = added.AId;
	}
}

public class C
{
	public Guid Id { get; set; }
	public Guid BId { get; set; }
	public string Name { get; set; }

	public void Apply(CAdded added)
	{
		Id = added.Id;
		BId = added.BId;
		Name = added.Name;
	}

	public void Apply(CNameChanged nameChanged)
	{
		Name = nameChanged.Name;
	}
}

// Events

public class AAdded
{
	public Guid Id { get; set; }
	public string Name { get; set; }
}

public class BAdded
{
	public Guid Id { get; set; }
	public Guid AId { get; set; }
}

public class CAdded
{
	public Guid Id { get; set; }
	public Guid BId { get; set; }
	public string Name { get; set; }
}

public class CNameChanged
{
	public string Name { get; set; }
}

// Projections

public class CNamesByA
{
	public Guid Id { get; set; }
	public Dictionary<Guid, string> CNamesByC { get; set; } = new Dictionary<Guid, string>();
}

public class CNamesByAProjection : Projection<CNamesByA>
{
	
	public CNamesByAProjection()
	{		
		Apply<CAdded>(
			(evt, view) => view.CNamesByC[evt.Data.Id] = evt.Data.Name,
			(session, stream, evt) => session.Load<B>(evt.Data.BId).AId);
		Apply<CNameChanged>(
			(evt, view) => view.CNamesByC[evt.StreamId] = evt.Data.Name,
			(session, stream, evt) => session.Load<B>(session.Load<C>(evt.StreamId).BId).AId);
	}
}

// IProjection implementation

public class Projection<TView> : IProjection where TView : class, new()
{
	private readonly IDictionary<Type, Transform> _transforms
		= new ConcurrentDictionary<Type, Transform>();

	public Type Produces => typeof(TView);

	public Type[] Consumes => _transforms.Keys.ToArray();

	public AsyncOptions AsyncOptions { get; } = new AsyncOptions();

	public Projection<TView> Apply<TEvent>(Action<Event<TEvent>, TView> transformation, Func<IDocumentSession, EventStream, Event<TEvent>, Guid> getViewId)
	{
		if (transformation == null)
		{
			throw new ArgumentException(nameof(transformation));
		}
		if (getViewId == null)
		{
			throw new ArgumentException(nameof(getViewId));
		}

		_transforms.Add(typeof(TEvent), new Transform
		{
			Transformation = (evt, view) => transformation((Event<TEvent>)evt, view),
			GetViewId = (session, stream, evt) => getViewId(session, stream, (Event<TEvent>)evt)
		});

		return this;
	}

	public Projection<TView> Apply<TEvent>(Action<Event<TEvent>, TView> transformation, Func<EventStream, Event<TEvent>, Guid> getViewId) =>
		Apply(transformation, (session, stream, evt) => getViewId(stream, evt));

	public Projection<TView> Apply<TEvent>(Action<Event<TEvent>, TView> transformation) =>
		Apply(transformation, (session, stream, evt) => stream.Id);

	void IProjection.Apply(IDocumentSession session, EventStream[] streams)
	{
		// First get all the view IDs, caching the event ID to view ID map
		List<EventToTransform> eventsToTransform = GetEventsToTransform(session, streams);

		// Load all the views
		session.LoadMany<TView>(eventsToTransform.Select(x => x.ViewId).Distinct().ToArray());

		// Apply the events, creating new views when needed
		// Make sure to cache which views have already been added since we may operate on the same view more than once
		HashSet<Guid> viewIds = new HashSet<System.Guid>();
		foreach (EventToTransform eventToTransform in eventsToTransform)
		{
			TView view = session.Load<TView>(eventToTransform.ViewId)
				?? GetNewView(session, eventToTransform.ViewId);
			eventToTransform.Transform.Transformation(eventToTransform.Event, view);
			if (viewIds.Add(eventToTransform.ViewId))
			{
				session.Store(view);
			}
		}
	}

	async Task IProjection.ApplyAsync(IDocumentSession session, EventStream[] streams, CancellationToken token)
	{
		// First get all the view IDs, caching the event ID to view ID map
		List<EventToTransform> eventsToTransform = GetEventsToTransform(session, streams);

		// Load all the views
		await session.LoadManyAsync<TView>(token, eventsToTransform.Select(x => x.ViewId).Distinct().ToArray()).ConfigureAwait(false);

		// Apply the events, creating new views when needed
		// Make sure to cache which views have already been added since we may operate on the same view more than once
		HashSet<Guid> viewIds = new HashSet<System.Guid>();
		foreach (EventToTransform eventToTransform in eventsToTransform)
		{
			TView view = await session.LoadAsync<TView>(eventToTransform.ViewId, token).ConfigureAwait(false)
				?? GetNewView(session, eventToTransform.ViewId);
			eventToTransform.Transform.Transformation(eventToTransform.Event, view);
			if (viewIds.Add(eventToTransform.ViewId))
			{
				session.Store(view);
			}
		}
	}

	private List<EventToTransform> GetEventsToTransform(IDocumentSession session, EventStream[] streams)
	{
		List<EventToTransform> eventsToTransform = new List<EventToTransform>();
		foreach (EventStream stream in streams)
		{
			Transform transform = null;
			foreach (IEvent evt in stream.Events.Where(evt => _transforms.TryGetValue(evt.Data.GetType(), out transform)))
			{
				Guid viewId = transform.GetViewId(session, stream, evt);
				eventsToTransform.Add(new EventToTransform
				{
					Stream = stream,
					Event = evt,
					Transform = transform,
					ViewId = viewId
				});
			}
		}
		return eventsToTransform;
	}

	private TView GetNewView(IDocumentSession session, Guid id)
	{
		TView view = new TView();
		session.DocumentStore.Schema.IdAssignmentFor<TView>().Assign(view, id);
		return view;
	}

	private class EventToTransform
	{
		public EventStream Stream { get; set; }
		public IEvent Event { get; set; }
		public Transform Transform { get; set; }
		public Guid ViewId { get; set; }
	}

	private class Transform
	{
		public Action<IEvent, TView> Transformation { get; set; }
		public Func<IDocumentSession, EventStream, IEvent, Guid> GetViewId { get; set; }
	}
}
```
</details>

This is similar, though has also diverged a bit, from this PR. Some key differences:
* There is a single handler for each event instead of splitting into two
* This does two "passes", the first to get all the view IDs from the events and then to fetch all the views in bulk (which may or may not trigger loads from the database to get the correct view ID) and then to apply the events to the views

TL;DR: The main thing I'd request is that the view ID selector function in the PR be passed the `IDocumentSession` so that you can get a view that requires loading additional documents to figure out it's ID. The rest looks like it mostly coding style or implementation details.

I can also see where there's a potential to combine this code with the existing aggregate code. There's really no reason there should be multiple APIs for generating aggregates. One API that is general enough to support different cases (like using in-view `Apply()` methods vs. a fluent API) should cover it. Maybe I'll take a look at some consolidation strategies once this one is accepted.

**Comment by tim-cools** [comment]
@daveaglick sure, sound like a good use case. 

I also like how you end up with only one handler collection, I try to incorporate that too.
I didn't use Apply as method name, because it doesn't apply the event, but it configures how the event will be projected. Hence the ProjectEvent name. But I'm open for discussion..

@jeremydmiller
One scenario that is not covered yet is the case where you have multiple view types per projection. This is a better alternative to using projections with dependencies to other dependencies. 

Base Projection --> Dependent Projection

If you have projection (dependent) that depends on another projection (base) you have to replay both in case you change the base projection. This replay mechanism can become quit complicated if you have multiple dependencies in you views.

What you can do instead is let the projection store all necessary data in its own views. If you do so you only need to replay the single projection in case somethings changes. And you don't have the complicated dependencies. You do end up with more duplicated data in the db, but data storage is cheap anyway.

**Comment by tim-cools** [comment]
I updated the PR with the overload that accepts the DocumentSession for selecting the id as requested by @daveaglick . I also had to refactor the ViewProjection to support this. This is loosely based on daves work.

**Comment by daveaglick** [comment]
Looking good, thanks! I'll try swapping out mine for this in my own tests and see if I hit any snags. :thumbsup:

**Comment by daveaglick** [comment]
:sparkles: :shipit:
